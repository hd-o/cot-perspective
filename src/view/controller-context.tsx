import type { Controller } from '@/controller'
import { createContext, useContext } from 'react'

export const ControllerContext = createContext({} as Controller)

export function useController() {
  return useContext(ControllerContext)
}
