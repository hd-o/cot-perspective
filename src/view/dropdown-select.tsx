import type { PropsWithChildren, ReactNode } from 'react'
import { config } from '@/common/config'

type DropdownSelectProps = PropsWithChildren<{
  defaultValue: string
}>

export function DropdownSelect(props: DropdownSelectProps): ReactNode {
  return (
    <select
      defaultValue={props.defaultValue}
      className={`custom-select form-control ${config.dropdownClass}`}
    >
      {props.children}
    </select>
  )
}
