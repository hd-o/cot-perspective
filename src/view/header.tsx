import type { PropsWithChildren, ReactNode } from 'react'

export function Header(props: PropsWithChildren): ReactNode {
  return (
    <header className="blog-header py-3">
      <div className="row flex-nowrap justify-content-between align-items-center">
        {props.children}
      </div>
    </header>
  )
}
