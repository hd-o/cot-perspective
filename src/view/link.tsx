import { createContext, FC, HTMLProps } from 'react'

const Link: FC<HTMLProps<HTMLAnchorElement>> = (props) => (
  <a
    href={props.href}
    title={props.title}
    className='text-dark'
    target='_blank'
    rel='noopener noreferrer'
  >
    {props.children}
  </a>
)

export const LinkCtx = createContext(Link)
