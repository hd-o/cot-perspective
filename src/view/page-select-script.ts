import { config } from '@/model/config'

declare global {
  interface Window {
    // Separate namespace to facilitate testing (using mocks)
    cotperspective: {
      assignLocation: typeof window.location.assign
    }
  }
}

/**
 * Handles page navigation after a dropdown option is selected.
 * Note: Might be refactored in the future if the site needs to
 * run React for more complex components, but at the moment this
 * fulfills the requirement while being a statically rendered site.
 */
export const pageSelectScript = `
  window.cotperspective = {
    assignLocation: window.location.assign.bind(window.location)
  }
  document.getElementById('cotperspective').addEventListener('change', (event) => {
    if (event.target.className.includes('${config.dropdownClass}')) {
      window.cotperspective.assignLocation(event.target.value + '.html')
    }
  })
`
