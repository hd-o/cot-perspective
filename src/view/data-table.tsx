import type { AverageInput } from '@/controller/data'
import type { ReactNode } from 'react'
import { useController } from './controller-context'

export function DataTable(props: AverageInput): ReactNode {
  const { data } = useController()

  const getAverageWithIndex = data.getAverage(props)
  const longAverage = getAverageWithIndex(1)
  const shortAverage = getAverageWithIndex(2)
  const longPercentageAverage = getAverageWithIndex(5)
  const shortPercentageAverage = getAverageWithIndex(6)
  const netAverage = getAverageWithIndex(7)

  return (
    <table
      className="table table-bordered text-center"
      style={{ marginTop: 10 }}
    >
      <thead className="thead-dark">
        <tr>
          <th scope="col">DATE</th>
          <th scope="col">LONG</th>
          <th scope="col">SHORT</th>
          <th scope="col">CHANGE LONG</th>
          <th scope="col">CHANGE SHORT</th>
          <th scope="col">% LONG</th>
          <th scope="col">% SHORT</th>
          <th scope="col">NET POSITIONS</th>
        </tr>
      </thead>
      <tbody>
        {props.values.length >= props.averagePeriod && (
          <tr className="bg-light">
            <td title={`${props.averagePeriod} Period Simple Average`}>
              Average(
              {props.averagePeriod}
              )
            </td>
            <td>{longAverage}</td>
            <td>{shortAverage}</td>
            <td colSpan={2}></td>
            <td>
              {longPercentageAverage}
              {' '}
              %
            </td>
            <td>
              {shortPercentageAverage}
              {' '}
              %
            </td>
            <td>{netAverage}</td>
          </tr>
        )}
        {props.values.map((row, valueIndex) => (
          <tr key={valueIndex}>
            {row.map((column, index) => {
              // Add classname to make table cell red/green
              // depending if the value increased/decreased
              let className = ''

              if (
                props.values[valueIndex + 1] !== undefined
                && (index === 1 || index === 2 || index === 7)
              ) {
                const previousColumn = props.values[valueIndex + 1][index]
                if (previousColumn > column) {
                  className = 'column-negative'
                }
                if (previousColumn < column) {
                  className = 'column-positive'
                }
              }

              // Some values are blank (not available) in COT's CSV
              let text = Number.isNaN(column) ? 'Not Available' : column

              // Format numbers for better readability
              if (index >= 1 || index <= 4 || index === 7) {
                text = text.toLocaleString()
              }
              if (index === 5 || index === 6) {
                text = `${text} %`
              }

              return (
                <td className={className} key={index}>
                  {text}
                </td>
              )
            })}
          </tr>
        ))}
      </tbody>
    </table>
  )
}
