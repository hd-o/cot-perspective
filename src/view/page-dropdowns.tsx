import type { COTData, DropDownSelections, TraderCategories } from '@/model/types'
import type { ReactNode } from 'react'
import { useController } from './controller-context'
import { DropdownSelect } from './dropdown-select'
import { pageSelectScript } from './page-select-script'

export type PageDropdownsProps = DropDownSelections & {
  data: COTData
  exchanges: string[]
  markets: string[]
  traderCategories: TraderCategories
}

export function PageDropdowns(props: PageDropdownsProps): ReactNode {
  const { file: file } = useController()

  const defaultExchangeValue = file.getPageId({
    exchange: props.exchange,
    market: Object.keys(props.data[props.exchange])[0],
    traderCategory: props.traderCategory,
  })

  const defaultSelectValue = file.getPageId({
    exchange: props.exchange,
    market: props.market,
    traderCategory: props.traderCategory,
  })

  const exchangesDropdown = (
    <DropdownSelect defaultValue={defaultExchangeValue}>
      {props.exchanges.map(exchange => (
        <option
          key={exchange}
          value={`${file.getPageId({
            exchange,
            market: Object.keys(props.data[exchange])[0],
            traderCategory: props.traderCategory,
          })}`}
        >
          {exchange}
        </option>
      ))}
    </DropdownSelect>
  )

  const marketsDropdown = (
    <DropdownSelect defaultValue={defaultSelectValue}>
      {props.markets.map(market => (
        <option
          key={market}
          value={`${file.getPageId({
            exchange: props.exchange,
            market,
            traderCategory: props.traderCategory,
          })}`}
        >
          {market}
        </option>
      ))}
    </DropdownSelect>
  )

  const tradersDropdown = (
    <DropdownSelect defaultValue={defaultSelectValue}>
      {props.traderCategories.map(traderCategory => (
        <option
          key={traderCategory}
          value={`${file.getPageId({
            exchange: props.exchange,
            market: props.market,
            traderCategory,
          })}`}
        >
          {traderCategory}
        </option>
      ))}
    </DropdownSelect>
  )

  return (
    <form style={{ marginTop: 20 }}>
      <div className="row">
        <div className="col">
          <div className="form-group">
            <label>Exchange</label>
            {exchangesDropdown}
          </div>
        </div>
        <div className="col">
          <div className="form-group">
            <label>Market</label>
            {marketsDropdown}
          </div>
        </div>
        <div className="col">
          <div className="form-group">
            <label>Trader</label>
            {tradersDropdown}
          </div>
        </div>
      </div>
      {/* Enable page navigation after selecting a dropdown option */}
      <script dangerouslySetInnerHTML={{ __html: pageSelectScript }} />
    </form>
  )
}
