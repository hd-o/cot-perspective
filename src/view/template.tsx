import fs from 'fs'
import React, { ReactNode } from 'react'
import { DataTable, DataTableProps } from './data-table'
import { PageDropDowns, PageDropDownsProps } from './page-dropdowns'

const license = fs.readFileSync('./LICENSE').toString()

const Link = (props: React.HTMLProps<HTMLAnchorElement>) => (
  <a
    href={props.href}
    title={props.title}
    className='text-dark'
    target='_blank'
    rel='noopener noreferrer'
  >
    {props.children}
  </a>
)

interface UsefulLinkProps {
  link: string,
  title: string,
  channel: string
}

const UsefulLink = ({ link, title, channel }: UsefulLinkProps) => (
  <li className='useful-link'>
    <Link href={link} title={`${channel} - ${title}`}>
      <strong>{channel}</strong> - {title}
    </Link>
  </li>
)

const Header = ({ children }: { children: ReactNode }) => (
  <header className='blog-header py-3'>
    <div className='row flex-nowrap justify-content-between align-items-center'>
      {children}
    </div>
  </header>
)

interface TemplateProps {
  dropDownsData: PageDropDownsProps
  tableData: DataTableProps
}

export const Template = ({ tableData, dropDownsData }: TemplateProps): JSX.Element => (
  <React.StrictMode>
    <html lang='en'>
      <head>
        <meta charSet='utf-8' />
        <link rel="icon" href="favicon.ico" type="image/x-icon" />
        <title>COT Perspective • Delightful COT Reports</title>
        <meta name='description' content='Delightful COT Reports' />
        <meta name='author' content='Hadrian de Oliveira' />
        <meta property="og:type" content="website" />
        <meta property="og:site_name" content="COT Perspective" />
        <meta property="og:title" content="COT Perspective • Delightful COT Reports" />
        <meta property="og:description" content="Delightful COT Reports" />
        <meta property="og:image" content="https://cotperspective.com/preview.png" />
        <meta property="og:url" content="https://cotperspective.com" />
        <link rel='stylesheet' href='https://bootswatch.com/4/cosmo/bootstrap.min.css' />
        <link rel='stylesheet' href='styles.css' />
      </head>
      <body>
        <div id='cotperspective' className='container'>
          <Header>
            <div className='col-4 pt-1' />
            <div className='col-4 text-center'>
              <a className='blog-header-logo text-dark' href='/'>COT Perspective</a>
            </div>
            <div className='col-4 d-flex justify-content-end align-items-center' />
          </Header>
          <PageDropDowns {...dropDownsData} />
          <DataTable {...tableData} />
          <Header>
            <div className='col-12'>
              <h3>Useful Links</h3>
            </div>
          </Header>
          <div className='row mb-2' style={{ paddingTop: 20 }}>
            <ul>
              <UsefulLink
                link='https://github.com/hd-o/cotperspective'
                channel='Source Code'
                title='github.com/hd-o/cotperspective'
              />
              <UsefulLink
                link='https://www.cftc.gov/MarketReports/CommitmentsofTraders/index.htm'
                channel='CFTC'
                title='Commitments of Traders (COT) Reports Descriptions'
              />
              <UsefulLink
                link='https://www.youtube.com/watch?v=mjaWvU6tzrM'
                channel='Magic Trader FX'
                title='CFTC COT Report Understanding the Institutions'
              />
              <UsefulLink
                link='https://www.youtube.com/watch?v=2EpSaM2H540'
                channel='Transparent Fx Academy'
                title='How To Read The COMMITMENT OF TRADERS - COT Report'
              />
            </ul>
          </div>
          <div className='row mb-2' style={{ paddingTop: 20, paddingBottom: 30 }}>
            <div className='container text-justify'>
              <small>
                {license}
              </small>
            </div>
          </div>
        </div>
      </body>
    </html>
  </React.StrictMode>
)
