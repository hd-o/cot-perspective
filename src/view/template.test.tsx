import { config } from '@/common/config'
import { Controller } from '@/controller'
import { ControllerContext } from '@/view/controller-context'
import { Template } from '@/view/template'
import renderer from 'react-test-renderer'
import { cotData } from '../model/mocks'

it('template snapshot', async () => {
  const ctrl = new Controller()
  const marketsData = cotData[config.defaultSelections.exchange]
  const marketData = marketsData[config.defaultSelections.market]
  const template = (
    <ControllerContext value={ctrl}>
      <Template
        dropDownsData={{
          data: cotData,
          exchanges: ctrl.data.getSortedKeys(cotData),
          markets: ctrl.data.getSortedKeys(cotData[config.defaultSelections.exchange]),
          traderCategories: config.traderCategories,
          ...config.defaultSelections,
        }}
        tableData={{
          averagePeriod: config.averagePeriod,
          values: marketData.map(
            ctrl.data.processTableData(config.defaultSelections.traderCategory),
          ),
        }}
      />
    </ControllerContext>
  )
  expect(renderer.create(template).toJSON()).toMatchSnapshot()
})
