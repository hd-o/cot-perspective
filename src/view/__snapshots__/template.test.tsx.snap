// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template snapshot 1`] = `
<html
  lang="en"
>
  <head>
    <meta
      charSet="utf-8"
    />
    <link
      href="favicon.ico"
      rel="icon"
      type="image/x-icon"
    />
    <title>
      COT Perspective • Delightful COT Reports
    </title>
    <meta
      content="Delightful COT Reports"
      name="description"
    />
    <meta
      content="Hadrian de Oliveira"
      name="author"
    />
    <meta
      content="website"
      property="og:type"
    />
    <meta
      content="COT Perspective"
      property="og:site_name"
    />
    <meta
      content="COT Perspective • Delightful COT Reports"
      property="og:title"
    />
    <meta
      content="Delightful COT Reports"
      property="og:description"
    />
    <meta
      content="https://cotperspective.com/preview.png"
      property="og:image"
    />
    <meta
      content="https://cotperspective.com"
      property="og:url"
    />
    <link
      href="https://bootswatch.com/4/cosmo/bootstrap.min.css"
      rel="stylesheet"
    />
    <link
      href="styles.css"
      rel="stylesheet"
    />
  </head>
  <body>
    <div
      className="container"
      id="cotperspective"
    >
      <header
        className="blog-header py-3"
      >
        <div
          className="row flex-nowrap justify-content-between align-items-center"
        >
          <div
            className="col-4 pt-1"
          />
          <div
            className="col-4 text-center"
          >
            <a
              className="blog-header-logo text-dark"
              href="/"
            >
              COT Perspective
            </a>
          </div>
          <div
            className="col-4 d-flex justify-content-end align-items-center"
          />
        </div>
      </header>
      <form
        style={
          {
            "marginTop": 20,
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col"
          >
            <div
              className="form-group"
            >
              <label>
                Exchange
              </label>
              <select
                className="custom-select form-control js-page-dropdown-select"
                defaultValue="chicagomercantileexchange3monthsofrnoncommercial"
              >
                <option
                  value="chicagomercantileexchange3monthsofrnoncommercial"
                >
                  CHICAGO MERCANTILE EXCHANGE
                </option>
                <option
                  value="icefuturesususdollarindexnoncommercial"
                >
                  ICE FUTURES U.S.
                </option>
              </select>
            </div>
          </div>
          <div
            className="col"
          >
            <div
              className="form-group"
            >
              <label>
                Market
              </label>
              <select
                className="custom-select form-control js-page-dropdown-select"
                defaultValue="chicagomercantileexchangeeurofxnoncommercial"
              >
                <option
                  value="chicagomercantileexchange3monthsofrnoncommercial"
                >
                  3-MONTH SOFR
                </option>
                <option
                  value="chicagomercantileexchangeeurofxnoncommercial"
                >
                  EURO FX
                </option>
                <option
                  value="chicagomercantileexchangeeurofxbritishpoundxratenoncommercial"
                >
                  EURO FX/BRITISH POUND XRATE
                </option>
              </select>
            </div>
          </div>
          <div
            className="col"
          >
            <div
              className="form-group"
            >
              <label>
                Trader
              </label>
              <select
                className="custom-select form-control js-page-dropdown-select"
                defaultValue="chicagomercantileexchangeeurofxnoncommercial"
              >
                <option
                  value="chicagomercantileexchangeeurofxcommercial"
                >
                  Commercial
                </option>
                <option
                  value="chicagomercantileexchangeeurofxnoncommercial"
                >
                  Noncommercial
                </option>
              </select>
            </div>
          </div>
        </div>
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "
  window.cotperspective = {
    assignLocation: window.location.assign.bind(window.location)
  }
  document.getElementById('cotperspective').addEventListener('change', (event) => {
    if (event.target.className.includes('js-page-dropdown-select')) {
      window.cotperspective.assignLocation(event.target.value + '.html')
    }
  })
",
            }
          }
        />
      </form>
      <table
        className="table table-bordered text-center"
        style={
          {
            "marginTop": 10,
          }
        }
      >
        <thead
          className="thead-dark"
        >
          <tr>
            <th
              scope="col"
            >
              DATE
            </th>
            <th
              scope="col"
            >
              LONG
            </th>
            <th
              scope="col"
            >
              SHORT
            </th>
            <th
              scope="col"
            >
              CHANGE LONG
            </th>
            <th
              scope="col"
            >
              CHANGE SHORT
            </th>
            <th
              scope="col"
            >
              % LONG
            </th>
            <th
              scope="col"
            >
              % SHORT
            </th>
            <th
              scope="col"
            >
              NET POSITIONS
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="bg-light"
          >
            <td
              title="12 Period Simple Average"
            >
              Average(
              12
              )
            </td>
            <td>
              176,454
            </td>
            <td>
              85,405
            </td>
            <td
              colSpan={2}
            />
            <td>
              28
               %
            </td>
            <td>
              13
               %
            </td>
            <td>
              91,049
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-07-07
            </td>
            <td
              className="column-positive"
            >
              186,241
            </td>
            <td
              className="column-positive"
            >
              80,047
            </td>
            <td
              className=""
            >
              5,904
            </td>
            <td
              className=""
            >
              1,730
            </td>
            <td
              className=""
            >
              28.8 %
            </td>
            <td
              className=""
            >
              12.4 %
            </td>
            <td
              className="column-positive"
            >
              106,194
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-06-30
            </td>
            <td
              className="column-negative"
            >
              180,338
            </td>
            <td
              className="column-positive"
            >
              78,317
            </td>
            <td
              className=""
            >
              -10,929
            </td>
            <td
              className=""
            >
              6,403
            </td>
            <td
              className=""
            >
              28.0 %
            </td>
            <td
              className=""
            >
              12.2 %
            </td>
            <td
              className="column-negative"
            >
              102,021
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-06-23
            </td>
            <td
              className="column-positive"
            >
              191,267
            </td>
            <td
              className="column-positive"
            >
              71,914
            </td>
            <td
              className=""
            >
              3,162
            </td>
            <td
              className=""
            >
              2,053
            </td>
            <td
              className=""
            >
              29.5 %
            </td>
            <td
              className=""
            >
              11.1 %
            </td>
            <td
              className="column-positive"
            >
              119,353
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-06-16
            </td>
            <td
              className="column-positive"
            >
              188,105
            </td>
            <td
              className="column-negative"
            >
              69,861
            </td>
            <td
              className=""
            >
              1,305
            </td>
            <td
              className=""
            >
              -19,391
            </td>
            <td
              className=""
            >
              29.5 %
            </td>
            <td
              className=""
            >
              11.0 %
            </td>
            <td
              className="column-positive"
            >
              118,244
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-06-09
            </td>
            <td
              className="column-positive"
            >
              186,800
            </td>
            <td
              className="column-negative"
            >
              89,251
            </td>
            <td
              className=""
            >
              12,662
            </td>
            <td
              className=""
            >
              -2,579
            </td>
            <td
              className=""
            >
              28.5 %
            </td>
            <td
              className=""
            >
              13.6 %
            </td>
            <td
              className="column-positive"
            >
              97,549
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-06-02
            </td>
            <td
              className="column-negative"
            >
              174,138
            </td>
            <td
              className="column-negative"
            >
              91,831
            </td>
            <td
              className=""
            >
              -105
            </td>
            <td
              className=""
            >
              -5,732
            </td>
            <td
              className=""
            >
              26.9 %
            </td>
            <td
              className=""
            >
              14.2 %
            </td>
            <td
              className="column-positive"
            >
              82,307
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-05-26
            </td>
            <td
              className="column-positive"
            >
              174,243
            </td>
            <td
              className="column-positive"
            >
              97,562
            </td>
            <td
              className=""
            >
              7,524
            </td>
            <td
              className=""
            >
              3,817
            </td>
            <td
              className=""
            >
              28.6 %
            </td>
            <td
              className=""
            >
              16.0 %
            </td>
            <td
              className="column-positive"
            >
              76,681
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-05-19
            </td>
            <td
              className="column-negative"
            >
              166,719
            </td>
            <td
              className="column-positive"
            >
              93,745
            </td>
            <td
              className=""
            >
              -3,425
            </td>
            <td
              className=""
            >
              1,257
            </td>
            <td
              className=""
            >
              27.7 %
            </td>
            <td
              className=""
            >
              15.6 %
            </td>
            <td
              className="column-negative"
            >
              72,974
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-05-12
            </td>
            <td
              className="column-positive"
            >
              170,143
            </td>
            <td
              className="column-positive"
            >
              92,488
            </td>
            <td
              className=""
            >
              4,569
            </td>
            <td
              className=""
            >
              886
            </td>
            <td
              className=""
            >
              28.4 %
            </td>
            <td
              className=""
            >
              15.5 %
            </td>
            <td
              className="column-positive"
            >
              77,655
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-05-05
            </td>
            <td
              className="column-positive"
            >
              165,575
            </td>
            <td
              className="column-positive"
            >
              91,602
            </td>
            <td
              className=""
            >
              763
            </td>
            <td
              className=""
            >
              5,891
            </td>
            <td
              className=""
            >
              27.4 %
            </td>
            <td
              className=""
            >
              15.1 %
            </td>
            <td
              className="column-negative"
            >
              73,973
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-04-28
            </td>
            <td
              className="column-negative"
            >
              164,811
            </td>
            <td
              className="column-positive"
            >
              85,712
            </td>
            <td
              className=""
            >
              -4,256
            </td>
            <td
              className=""
            >
              3,182
            </td>
            <td
              className=""
            >
              27.1 %
            </td>
            <td
              className=""
            >
              14.1 %
            </td>
            <td
              className="column-negative"
            >
              79,099
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-04-21
            </td>
            <td
              className="column-positive"
            >
              169,068
            </td>
            <td
              className="column-positive"
            >
              82,530
            </td>
            <td
              className=""
            >
              5,425
            </td>
            <td
              className=""
            >
              4,086
            </td>
            <td
              className=""
            >
              27.8 %
            </td>
            <td
              className=""
            >
              13.6 %
            </td>
            <td
              className="column-positive"
            >
              86,538
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-04-14
            </td>
            <td
              className="column-positive"
            >
              163,643
            </td>
            <td
              className="column-negative"
            >
              78,444
            </td>
            <td
              className=""
            >
              2,440
            </td>
            <td
              className=""
            >
              -4,657
            </td>
            <td
              className=""
            >
              27.2 %
            </td>
            <td
              className=""
            >
              13.1 %
            </td>
            <td
              className="column-positive"
            >
              85,199
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-04-07
            </td>
            <td
              className="column-positive"
            >
              161,204
            </td>
            <td
              className="column-positive"
            >
              83,102
            </td>
            <td
              className=""
            >
              6,022
            </td>
            <td
              className=""
            >
              832
            </td>
            <td
              className=""
            >
              26.2 %
            </td>
            <td
              className=""
            >
              13.5 %
            </td>
            <td
              className="column-positive"
            >
              78,102
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-03-31
            </td>
            <td
              className="column-positive"
            >
              155,182
            </td>
            <td
              className="column-negative"
            >
              82,269
            </td>
            <td
              className=""
            >
              5,445
            </td>
            <td
              className=""
            >
              -11,733
            </td>
            <td
              className=""
            >
              24.4 %
            </td>
            <td
              className=""
            >
              13.0 %
            </td>
            <td
              className="column-positive"
            >
              72,913
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-03-24
            </td>
            <td
              className="column-positive"
            >
              149,736
            </td>
            <td
              className="column-negative"
            >
              94,002
            </td>
            <td
              className=""
            >
              1,583
            </td>
            <td
              className=""
            >
              -23,066
            </td>
            <td
              className=""
            >
              22.5 %
            </td>
            <td
              className=""
            >
              14.1 %
            </td>
            <td
              className="column-positive"
            >
              55,734
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-03-17
            </td>
            <td
              className="column-negative"
            >
              148,153
            </td>
            <td
              className="column-negative"
            >
              117,068
            </td>
            <td
              className=""
            >
              -4,126
            </td>
            <td
              className=""
            >
              -47,647
            </td>
            <td
              className=""
            >
              21.8 %
            </td>
            <td
              className=""
            >
              17.2 %
            </td>
            <td
              className="column-positive"
            >
              31,085
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-03-10
            </td>
            <td
              className="column-negative"
            >
              152,279
            </td>
            <td
              className="column-negative"
            >
              164,716
            </td>
            <td
              className=""
            >
              -868
            </td>
            <td
              className=""
            >
              -72,432
            </td>
            <td
              className=""
            >
              20.2 %
            </td>
            <td
              className=""
            >
              21.8 %
            </td>
            <td
              className="column-positive"
            >
              -12,437
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-03-03
            </td>
            <td
              className="column-negative"
            >
              153,147
            </td>
            <td
              className="column-negative"
            >
              237,148
            </td>
            <td
              className=""
            >
              -324
            </td>
            <td
              className=""
            >
              -40,089
            </td>
            <td
              className=""
            >
              20.4 %
            </td>
            <td
              className=""
            >
              31.5 %
            </td>
            <td
              className="column-positive"
            >
              -84,001
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-02-25
            </td>
            <td
              className="column-negative"
            >
              153,472
            </td>
            <td
              className="column-positive"
            >
              277,237
            </td>
            <td
              className=""
            >
              -15,222
            </td>
            <td
              className=""
            >
              6,681
            </td>
            <td
              className=""
            >
              21.6 %
            </td>
            <td
              className=""
            >
              39.0 %
            </td>
            <td
              className="column-negative"
            >
              -123,765
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-02-18
            </td>
            <td
              className="column-positive"
            >
              168,693
            </td>
            <td
              className="column-positive"
            >
              270,555
            </td>
            <td
              className=""
            >
              313
            </td>
            <td
              className=""
            >
              10,686
            </td>
            <td
              className=""
            >
              24.2 %
            </td>
            <td
              className=""
            >
              38.8 %
            </td>
            <td
              className="column-negative"
            >
              -101,862
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-02-11
            </td>
            <td
              className="column-positive"
            >
              168,381
            </td>
            <td
              className="column-positive"
            >
              259,869
            </td>
            <td
              className=""
            >
              682
            </td>
            <td
              className=""
            >
              12,193
            </td>
            <td
              className=""
            >
              25.0 %
            </td>
            <td
              className=""
            >
              38.6 %
            </td>
            <td
              className="column-negative"
            >
              -91,488
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-02-04
            </td>
            <td
              className="column-negative"
            >
              167,699
            </td>
            <td
              className="column-positive"
            >
              247,676
            </td>
            <td
              className=""
            >
              -10,186
            </td>
            <td
              className=""
            >
              3,693
            </td>
            <td
              className=""
            >
              25.7 %
            </td>
            <td
              className=""
            >
              38.0 %
            </td>
            <td
              className="column-negative"
            >
              -79,977
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-01-28
            </td>
            <td
              className="column-positive"
            >
              177,885
            </td>
            <td
              className="column-positive"
            >
              243,983
            </td>
            <td
              className=""
            >
              12,606
            </td>
            <td
              className=""
            >
              26,679
            </td>
            <td
              className=""
            >
              27.2 %
            </td>
            <td
              className=""
            >
              37.2 %
            </td>
            <td
              className="column-negative"
            >
              -66,098
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-01-21
            </td>
            <td
              className="column-positive"
            >
              165,279
            </td>
            <td
              className="column-positive"
            >
              217,304
            </td>
            <td
              className=""
            >
              3,396
            </td>
            <td
              className=""
            >
              2,873
            </td>
            <td
              className=""
            >
              26.3 %
            </td>
            <td
              className=""
            >
              34.6 %
            </td>
            <td
              className="column-positive"
            >
              -52,025
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-01-14
            </td>
            <td
              className="column-negative"
            >
              161,883
            </td>
            <td
              className="column-negative"
            >
              214,431
            </td>
            <td
              className=""
            >
              -2,134
            </td>
            <td
              className=""
            >
              -14,275
            </td>
            <td
              className=""
            >
              26.3 %
            </td>
            <td
              className=""
            >
              34.9 %
            </td>
            <td
              className="column-positive"
            >
              -52,548
            </td>
          </tr>
          <tr>
            <td
              className=""
            >
              2020-01-07
            </td>
            <td
              className=""
            >
              164,017
            </td>
            <td
              className=""
            >
              228,706
            </td>
            <td
              className=""
            >
              -1,827
            </td>
            <td
              className=""
            >
              -12,365
            </td>
            <td
              className=""
            >
              27.5 %
            </td>
            <td
              className=""
            >
              38.3 %
            </td>
            <td
              className=""
            >
              -64,689
            </td>
          </tr>
        </tbody>
      </table>
      <header
        className="blog-header py-3"
      >
        <div
          className="row flex-nowrap justify-content-between align-items-center"
        >
          <div
            className="col-12"
          >
            <h3>
              Useful Links
            </h3>
          </div>
        </div>
      </header>
      <div
        className="row mb-2"
        style={
          {
            "paddingTop": 20,
          }
        }
      >
        <ul>
          <li
            className="useful-link"
          >
            <a
              className="text-dark"
              href="https://github.com/hd-o/cotperspective"
              rel="noopener noreferrer"
              target="_blank"
              title="Source Code - github.com/hd-o/cotperspective"
            >
              <strong>
                Source Code
              </strong>
               - 
              github.com/hd-o/cotperspective
            </a>
          </li>
          <li
            className="useful-link"
          >
            <a
              className="text-dark"
              href="https://www.cftc.gov/MarketReports/CommitmentsofTraders/index.htm"
              rel="noopener noreferrer"
              target="_blank"
              title="CFTC - Commitments of Traders (COT) Reports Descriptions"
            >
              <strong>
                CFTC
              </strong>
               - 
              Commitments of Traders (COT) Reports Descriptions
            </a>
          </li>
          <li
            className="useful-link"
          >
            <a
              className="text-dark"
              href="https://www.youtube.com/watch?v=mjaWvU6tzrM"
              rel="noopener noreferrer"
              target="_blank"
              title="Magic Trader FX - CFTC COT Report Understanding the Institutions"
            >
              <strong>
                Magic Trader FX
              </strong>
               - 
              CFTC COT Report Understanding the Institutions
            </a>
          </li>
          <li
            className="useful-link"
          >
            <a
              className="text-dark"
              href="https://www.youtube.com/watch?v=2EpSaM2H540"
              rel="noopener noreferrer"
              target="_blank"
              title="Transparent Fx Academy - How To Read The COMMITMENT OF TRADERS - COT Report"
            >
              <strong>
                Transparent Fx Academy
              </strong>
               - 
              How To Read The COMMITMENT OF TRADERS - COT Report
            </a>
          </li>
        </ul>
      </div>
      <div
        className="row mb-2"
        style={
          {
            "paddingBottom": 30,
            "paddingTop": 20,
          }
        }
      >
        <div
          className="container text-justify"
        >
          <small>
            MIT License

Copyright (c) Hadrian de Oliveira.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

          </small>
        </div>
      </div>
    </div>
  </body>
</html>
`;
